infra/cache/kInfraRedisCacheWrapper.php

```diff
@@ -0,0 +1,314 @@
+						break;
+
+					case self::STAT_OP:
+						KalturaMonitorClient::monitorRedisAccess($statsKey, $stats[self::STAT_TIME], $stats[self::STAT_COUNT]);
+						break;
+				}
+			}
+		}
+
+		self::$_stats = array();
+	}
+
+	public static function outputStats()
+	{
+		foreach (self::$_stats as $statsKey => $typeStats)
+		{
+			$cur = 'instance:' . $statsKey;
+			foreach ($typeStats as $type => $stats)
+			{
+				foreach ($stats as $key => $value)
+				{
+					$cur .= ', ' . $type . '_' . $key . ':' . $value;
+				}
+			}
+
+			KalturaLog::log($cur);
+		}
+	}
+
+	protected function updateStats($type, $stats)
+	{
+		if (!isset(self::$_stats[$this->statsKey]))
+		{
+			self::$_stats[$this->statsKey] = array();
+		}
+		$typeStats = &self::$_stats[$this->statsKey];
+
+		if (!isset($typeStats[$type]))
+		{
+			$typeStats[$type] = array();
+		}
+
+		foreach ($stats as $key => $value)
+		{
+			if (isset($typeStats[$type][$key]))
+			{
+				$typeStats[$type][$key] += $value;
+			}
+			else
+			{
+				$typeStats[$type][$key] = $value;
+			}
+		}
+	}
+
+	/**
+	 * @param int $errno
+	 * @param string $errstr
+	 * @return bool
+	 */
+	public function errorHandler($errno, $errstr)
+	{
+		self::safeLog("got error from redis [$errno] [$errstr]");
+		$this->gotError = true;
+		return false;
+	}
+
+	protected function callAndDetectErrors($methodName, $params)
+	{
+		while ($this->redis)
+		{
+			$this->gotError = false;
+
+			set_error_handler(array($this, 'errorHandler'));
+			$start = microtime(true);
+			$result = call_user_func_array(array($this->redis, $methodName), $params);
+			$end = microtime(true);
+			restore_error_handler();
+
+			$this->updateStats(self::STAT_OP, array(
+				self::STAT_COUNT => 1,
+				self::STAT_TIME => $end - $start));
+
+			if (!$this->gotError)
+			{
+				return $result;
+			}
+			$this->reconnect();
+		}
+
+		self::safeLog("There is no active redis connection");
+		return false;
+	}
+
+	/**
+	 * @return bool false on error
+	 */
+	protected function reconnect()
+	{
+		$this->close();
+
+		if ($this->connectAttempts >= self::MAX_CONNECT_ATTEMPTS)
+		{
+			return false;
+		}
+
+		$connectResult = false;
+		$connStart = microtime(true);
+		if(!$this->cluster)
+		{
+			while ($this->connectAttempts < self::MAX_CONNECT_ATTEMPTS)
+			{
+				$redis = new Redis();
+				$curConnStart = microtime(true);
+				try
+				{
+					if ($this->persistent)
+						$redis->pconnect($this->hostName, $this->port, $this->timeout);
+					else
+						$redis->connect($this->hostName, $this->port, $this->timeout);
+				} catch (Exception $e)
+				{
+					self::safeLog("failed to connect to redis");
+				}
+
+				$connectResult = $redis->isConnected();
+				if ($connectResult || microtime(true) - $curConnStart < .5)        // retry only if there's an error and it's a timeout error
+					break;
+
+
+				self::safeLog("got timeout error while connecting to redis...");
+			}
+		}
+
+		else
+		{
+			try
+			{
+				if ($this->persistent)
```

Comment:
The creation does not match method signature:
redis_cluster_init(redisCluster *c, HashTable *ht_seeds, double timeout,
                               double read_timeout, int persistent, zend_string *user,
                               zend_string *pass, zval *context)
                               
Also better consolidate to one while sending the persistent value as param.